framework/rediscache.py

```diff
@@ -0,0 +1,109 @@
+
+
+# -*- coding: utf-8 -*-
+# Copyright 2020 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License")
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import json
+import logging
+
+from google.cloud import ndb
+import redis
+import fakeredis
+
+
+UNIT_TEST_MODE = os.environ['SERVER_SOFTWARE'].startswith('test')
+if not UNIT_TEST_MODE:
+  # Create a Redis client.
+  redis_host = os.environ.get('REDISHOST', 'localhost')
+  redis_port = int(os.environ.get('REDISPORT', 6379))
+  redis_client = redis.Redis(host=redis_host, port=redis_port, decode_responses=True)
+else:
+  redis_client = fakeredis.FakeStrictRedis(decode_responses=True)
+
+
+def set(key, value, time=None):
+  """
+  Redis SET sets the str key, value pair, https://redis.io/commands/set/; if
+  ``key`` already holds a value, it is overwritten.
+
+  ``time`` sets the expire time for this key, in seconds.
+  """
+  if not isinstance(value, str):
+    logging.info(
+        'value %s is not an instance of str, abort set caching', value)
+    return
+
+  if time:
+    redis_client.set(key, value, ex=time)
```

Comment:
I would like it to be safe to deploy a new version of the app or access non-live versions of the app.  One risk is that a shared cache could cause cross-talk between instances running different versions of the app.  A way to avoid that would be to prefix the key with the GAE app version.  That can be found in env variable GAE_VERSION.

Reply:
Also per offline discussions, we should set expire date all for all cache to avoid cache buildup overtime. What about 24 hours?

Reply:
24 hours seems fine

Reply:
Done